### DISCUSSION TITLE
Reproducible releases to Prod - Terragrunt

### DISCUSSION THREAD

**Discussion by CalvinRodo** [cds-snc/docs#50] [General]
I'm looking at writing an RFC defining Terragrunt as our standard way of managing IaC. 

So in thinking about this I'm trying to figure out a the best way to manage reproducible production releases.

At the moment Notify and EN Server Metrics both use a similar way of releasing to production. 

We tag the code base and we reference that tag in our terragrunt HCL files.

See an example here: https://github.com/cds-snc/covid-alert-metrics-terraform/blob/017eb96418411f41179a37283caf416e2fc44d97/env/production/s3/terragrunt.hcl#L7

When we do a release to production: 
https://github.com/cds-snc/covid-alert-metrics-terraform/blob/main/.github/workflows/apply-production.yml
https://github.com/cds-snc/notification-terraform/blob/main/.github/workflows/merge_to_main_production.yml

We get the latest version of the code that is being merged to main and run an apply on that. 

This means that our terragrunt configuration (stored in the `env` folder) at the time of applying to production is what is on main at that point and our terraform configuration (stored in the `aws` folder) is what is tagged. 

This is fine as long as that is our intention and means that if we want to reproduce a production release we need to checkout the codebase at the time of the PR merge and that we can't just set the manifest to the version we want to release because we will get the latest terragrunt configuration. 

So what do we want to set as the standard for CDS infra.

Option 1: 
  - Tagged Terraform Code
  - Main branch Terragrunt Configuration
  
Option 2: 
 - Tagged Terraform Code 
 - Tagged Terragrunt Configuration

**Comment by dj2** [comment]
Option 1 to me seems like it will lead to confusion as the used configuration doesn't match what you see in the tag, yea? The Option 2 seems safer as the tag is the only thing you need to re-release.

**Comment by patheard** [comment]
I like option 2 as well since the behaviour is more expected to me.  To repeat the build, just checkout that tag instead of needing to find the commit SHA for main at a given point in time.
